File: NewsApp/app/src/main/java/com/example/android/topic41/presentation/viewmodel/ViewModelProviderFactory.java
package com.example.android.topic41.presentation.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.android.topic41.domain.usecase.SingleUseCaseInterface;

public class ViewModelProviderFactory implements ViewModelProvider.Factory {
    SingleUseCaseInterface useCase;

    public ViewModelProviderFactory(SingleUseCaseInterface useCase) {
        super();
        this.useCase = useCase;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass == ArticlesViewModel.class) {
            return (T) new ArticlesViewModel(useCase);
        }
        return null;
    }
}

File: NewsApp/app/src/main/java/com/example/android/topic41/presentation/viewmodel/ViewModelProviderFactory.kt
package com.example.android.topic41.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.android.topic41.domain.usecase.SingleUseCaseInterface

class ViewModelProviderFactory(var useCase: SingleUseCaseInterface) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return when(modelClass) {
            ArticlesViewModel::class.java -> ArticlesViewModel(useCase) as T
            else -> throw Exception("Class is not MainViewModel")
        }
    }
}